services/QuillLMS/spec/controllers/teachers/progress_reports/diagnostic_reports_controller_spec.rb

```diff
@@ -515,4 +515,114 @@
       expect(classroom_unit2.reload.assigned_student_ids).to contain_exactly(student1.id, student2.id)
     end
   end
+
+  describe '#individual_student_diagnostic_responses' do
+    let(:teacher)       { create(:teacher) }
+    let(:student)       { create(:student) }
+    let(:classroom)     { create(:classroom) }
+    let(:unit)          { create(:unit) }
+    let(:activity1)    { create(:diagnostic_activity) }
+    let(:activity2)    { create(:diagnostic_activity) }
+    let!(:classroom_unit) { create(:classroom_unit, classroom:, unit:) }
+
+    context 'when no activity session exists' do
+      it 'returns a 404 status with empty JSON data' do
+        get :individual_student_diagnostic_responses, params: {
+          activity_id: activity1.id,
+          classroom_id: classroom.id,
+          student_id: student.id,
+          unit_id: unit.id
+        }
+        expect(response.status).to eq(404)
+        json = JSON.parse(response.body)
+        expect(json).to eq({})
+      end
+    end
+
+    context 'when an activity session exists' do
+      let!(:activity_session1) do
+        create(:activity_session,
+          activity: activity1,
+          classroom_unit:,
+          state: 'finished',
+          user: student)
+      end
+
+      before do
+        # Stub out methods that format results so we can predict the JSON structure.
+        allow(controller).to receive(:format_concept_results) do |session, _results_relation|
+          session == activity_session1 ? ['session1_concept'] : ['unknown']
```

Comment:
Is this ternary necessary?  It looks like there are only tests that go through one branch.

Reply:
It just helps confirm that the right session is getting passed to `format_concept_results`, but maybe just returning null otherwise is cleaner (and will mirror the logic before, where I think you're right that we don't specifically need `unknown`.